basis/math/matrices/matrices.factor

```diff
@@ -200,77 +121,155 @@ ERROR: negative-power-matrix m n ;
-    ] if-empty ;
+: vdotm ( v m -- p ) flip [ vdot ] with map ;
+: mdotv ( m v -- p ) [ vdot ] curry map ;
+: mdot ( m m -- m ) flip [ swap mdotv ] curry map ;
 
-: dim ( matrix -- pair/f )
-    [ 2 0 <array> ]
-    [ [ length ] [ first length ] bi 2array ] if-empty ;
+: m~  ( m1 m2 epsilon -- ? ) [ v~ ] curry 2all? ;
 
-: square-matrix? ( matrix -- ? )
-    { [ well-formed-matrix? ] [ dim all-eq? ] } 1&& ;
+: mmin ( m -- n ) [ 1/0. ] dip [ [ min ] each ] each ;
+: mmax ( m -- n ) [ -1/0. ] dip [ [ max ] each ] each ;
+: mnorm ( m -- m' ) dup mmax abs m/n ;
+: m-infinity-norm ( m -- n ) [ [ abs ] map-sum ] map supremum ;
+: m-1norm ( m -- n ) flip m-infinity-norm ;
+: frobenius-norm ( m -- n ) [ [ sq ] map-sum ] map-sum sqrt ;
 
-: matrix-coordinates ( dim -- coordinates )
-    first2 [ <iota> ] bi@ cartesian-product ; inline
+! well-defined for square matrices; but works on nonsquare too
+: main-diagonal ( matrix -- seq )
+    >square-matrix [ swap nth ] map-index ; inline
 
-: dimension-range ( matrix -- dim range )
-    dim [ matrix-coordinates ] [ first [1,b] ] bi ;
+! top right to bottom left; reverse the result if you expected it to start in the lower left
+: anti-diagonal ( matrix -- seq )
+    >square-matrix [ swap nth-end ] map-index ; inline
 
-: upper-matrix-indices ( matrix -- matrix' )
-    dimension-range <reversed> [ tail-slice* >array ] 2map concat ;
+<PRIVATE
+: (rows-iota) ( matrix -- rows-iota )
+    dim first <iota> ;
+: (cols-iota) ( matrix -- cols-iota )
+    dim second <iota> ;
 
-: lower-matrix-indices ( matrix -- matrix' )
-    dimension-range [ head-slice >array ] 2map concat ;
+: simple-rows-except ( matrix desc quot -- others )
+    curry [ dup (rows-iota) ] dip
+    pick reject-as swap rows ; inline
+
+: simple-cols-except ( matrix desc quot -- others )
+    curry [ dup (cols-iota) ] dip
+    pick reject-as swap cols transpose ; inline ! need to un-transpose the result of cols
+
+CONSTANT: scalar-except-quot [ = ]
+CONSTANT: sequence-except-quot [ member? ]
+PRIVATE>
 
-: make-lower-matrix ( object m n -- matrix )
-    zero-matrix [ lower-matrix-indices ] [ set-indices ] [ ] tri ;
+GENERIC: rows-except ( matrix desc -- others )
+M: integer rows-except  scalar-except-quot   simple-rows-except ;
+M: sequence rows-except sequence-except-quot simple-rows-except ;
 
-: make-upper-matrix ( object m n -- matrix )
-    zero-matrix [ upper-matrix-indices ] [ set-indices ] [ ] tri ;
+GENERIC: cols-except ( matrix desc -- others )
+M: integer cols-except  scalar-except-quot   simple-cols-except ;
+M: sequence cols-except sequence-except-quot simple-cols-except ;
+
+! well-defined for any well-formed-matrix
+: matrix-except ( matrix exclude-pair -- submatrix )
+    first2 [ rows-except ] dip cols-except ;
+
+:: matrix-except-all ( matrix-seq -- expansion )
+    matrix-seq dim [ <iota> ] map first2 cartesian-product
+    [ [ matrix-seq swap matrix-except ] map ] map ;
+
+: dim ( matrix -- dimensions )
```

Comment:
Probably should be ``matrix-dim`` for now, we have a lot of generic ``dim`` things?

Reply:
sounds good